### PR TITLE
feat: Introduce evaluation UI and PDF report generation

### PR DESCRIPTION
**Pull request by GrayNekoBean**
## Purpose

Add a frontend page for evaluation and integrate the compare model feature into report generation. 

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

- [ ] Yes
- [x] No

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

- [ ] Yes
- [x] No

## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [x] The current tests all pass (`python -m pytest`).
- [ ] I added tests that prove my fix is effective or that my feature works
- [ ] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [ ] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

### CODE REVIEW

**Review by phoevos — CHANGES_REQUESTED** [review]
I still haven't had the chance to properly look through the PR, but until I do I would really like to see issues I've mentioned before fixed. The main one is discrepancies between the PR code and `main`. If something doesn't need to change, it shouldn't. Please take a close look at the code (especially in cli.py, and evaluate.py) and the evaluation README.

There are minor stuff like the mistake in the `--num-questions` CLI args in `.vsode/launch.json`, as well as things I've bitched about in literally every PR before, like:
* Every function/method should have a docstring
* Docstrings should follow the [Google styleguide conventions](https://google.github.io/styleguide/pyguide.html#381-docstrings) like starting with a capital, ending with a period, no newline between them and the code, as well as general conventions like being formulated in the imperative
* There should be no commented-out code left behind
* The CI should be passing. Just merged some changes from main, fingers crossed it will improve the situation

Also, mentioning again that we're not using the `load_env` scripts and that the format of `summary.json` shouldn't change.

Apart from that, I noticed during the demo that the header of our page ("Chat" "Ask a Question", "Evaluate") is not evenly spaced. Although I won't go into too much depth on the frontend side, I do believe that this one should be fixed.

Finally, did you handle properly generating the environment needed for the evaluation to run? Are you running `create_eval_dotenv`? Are you fetching the values in a different way? How are they being passed to the evaluation script?

Regardless of all of the above, have you tested a fresh deployment?

Given that the PR was opened 3 days before the submission deadline it's unlikely that we'll be able to do a proper review. But I do expect that all superficial issues are resolved. Also just to be crystal clear: there is extreme code duplication for the "evaluation by request" feature and code has been moved around to facilitate just that (e.g. constants being moved into the evaluate module although they clearly don't belong there, a new function in evaluate running red teaming stuff- not the place to put it most likely- let alone repeating the whole logic and configuration parsing done in the existing functions meant to run the evaluation). This is ok if writing code that needs to be demoed in 2 hours, but not quite PR material. I kindly ask that we push to improve this PR as much as possible before the deadline.

**Comment by GrayNekoBean** [comment]
> there is extreme code duplication for the "evaluation by request" feature and code has been moved around to facilitate just that (e.g. constants being moved into the evaluate module although they clearly don't belong there, a new function in evaluate running red teaming stuff- not the place to put it most likely- let alone repeating the whole logic and configuration parsing done in the existing functions meant to run the evaluation).

Thanks, I also just noticed that those two functions' behaviour was more similar than I thought like an hour ago and already removed one of them. (more specifically, combined them) Removing that should also rise the coverage.

Me and binxuan will get these fixed asap, but we'll also keep the changes where we see necessary.

**Comment by binxuan39** [comment]
> I still haven't had the chance to properly look through the PR, but until I do I would really like to see issues I've mentioned before fixed. The main one is discrepancies between the PR code and `main`. If something doesn't need to change, it shouldn't. Please take a close look at the code (especially in cli.py, and evaluate.py) and the evaluation README.
> 
> There are minor stuff like the mistake in the `--num-questions` CLI args in `.vsode/launch.json`, as well as things I've bitched about in literally every PR before, like:
> 
> * Every function/method should have a docstring
> * Docstrings should follow the [Google styleguide conventions](https://google.github.io/styleguide/pyguide.html#381-docstrings) like starting with a capital, ending with a period, no newline between them and the code, as well as general conventions like being formulated in the imperative
> * There should be no commented-out code left behind
> * The CI should be passing. Just merged some changes from main, fingers crossed it will improve the situation
> 
> Also, mentioning again that we're not using the `load_env` scripts and that the format of `summary.json` shouldn't change.
> 
> Apart from that, I noticed during the demo that the header of our page ("Chat" "Ask a Question", "Evaluate") is not evenly spaced. Although I won't go into too much depth on the frontend side, I do believe that this one should be fixed.
> 
> Finally, did you handle properly generating the environment needed for the evaluation to run? Are you running `create_eval_dotenv`? Are you fetching the values in a different way? How are they being passed to the evaluation script?
> 
> Regardless of all of the above, have you tested a fresh deployment?
> 
> Given that the PR was opened 3 days before the submission deadline it's unlikely that we'll be able to do a proper review. But I do expect that all superficial issues are resolved. Also just to be crystal clear: there is extreme code duplication for the "evaluation by request" feature and code has been moved around to facilitate just that (e.g. constants being moved into the evaluate module although they clearly don't belong there, a new function in evaluate running red teaming stuff- not the place to put it most likely- let alone repeating the whole logic and configuration parsing done in the existing functions meant to run the evaluation). This is ok if writing code that needs to be demoed in 2 hours, but not quite PR material. I kindly ask that we push to improve this PR as much as possible before the deadline.

**Comment by binxuan39** [comment]
> There are minor stuff like the mistake in the --num-questions CLI args in .vsode/launch.json, as well as things I've bitched about in literally every PR before, like:

The correct one should be connected like we using now as the Typer automatically converts the underline to the hyphen.

File: app/backend/app.py [hunk 1]
```diff
@@ -339,6 +360,97 @@ async def speech():
...
+
+        if input_data_file is not None:
+            input_data = [json.loads(line) for line in input_data_file.readlines()]
+        save_jsonl(input_data, Path("./evaluation/input/input_temp.jsonl"))
+        config["testdata_path"] = "input/input_temp.jsonl"
+        config["results_dir"] = "results"
+
+        temp_config_path = Path("config_temp.json")
+
+        save_config(config, "evaluation" / temp_config_path)
```
Comment (phoevos) [inline app/backend/app.py]: Looks like you would be overwriting the same config file if more than one requests come in. On top of that, it doesn't look like you ever delete it.

**Review by phoevos — CHANGES_REQUESTED** [review]

File: app/backend/app.py (hunk 1 above)
Comment (GrayNekoBean) [reply app/backend/app.py]: It looks like the temp file will be loaded instantly after been written so I guess it doesn't really affect the running result in normal situations (these lines look a bit dumb though). They should be some sort of legacy code, I'll handle it.

File: .github/workflows/model-evaluation.yml
Comment (phoevos) [inline .github/workflows/model-evaluation.yml:1]: We need to add the report generation functionality here (as well as the corresponding artifact.)

**Review by phoevos — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 2]
```diff
@@ -153,6 +146,17 @@ npm-debug.log*
 node_modules
 static/
 
+# Hash files
 data/**/*.md5
 
+# Evaluation
+app/backend/evaluation/report
```
Comment (phoevos) [inline .gitignore]: Is anything ever generated in there? AFAICT, unless we explicitly specify the path, this should go under the timestamped experiment directories.

**Review by phoevos — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
Comment (GrayNekoBean) [reply .gitignore]: This is only used in some local machine tests, but its okay to remove that

**Comment by phoevos** [comment]
I'm not sure about the whole process of adding the red teaming in the evaluation bit. The combined process was always supposed to be a separate command that runs both and generates the report. I was expecting a new function that calls what we already had there in the evaluation and red teaming modules instead of patching up the evaluate part to do some extra things it wasn't intended to do. I'll try to do this refactoring myself, cause this keeps going down the wrong path.

PS: Deprecating the red-teaming command and only exposing a single evaluation that takes care of everything is absolutely an option, but it was never discussed.

**Comment by GrayNekoBean** [comment]
> I'm not sure about the whole process of adding the red teaming in the evaluation bit. The combined process was always supposed to be a separate command that runs both and generates the report. I was expecting a new function that calls what we already had there in the evaluation and red teaming modules instead of patching up the evaluate part to do some extra things it wasn't intended to do. I'll try to do this refactoring myself, cause this keeps going down the wrong path.
> 
> PS: Deprecating the red-teaming command and only exposing a single evaluation that takes care of everything is absolutely an option, but it was never discussed.

I totally get what you said, ideally I should be making a new function that invokes both parts and I was trying to do that in the beginning.  But that means I need to move report generation lines to an upper layer of the function call, and I was afraid that could make our code look more nested. I could do the refactor if the services are all good, but now I don't dare to make huge changes without testing it.

**Review by phoevos — APPROVED** [review]

**Pull request merged**